openmsupply/open-msupply
# 66 initialisation phase

closes: #66 

Soo many changes, there are heaps of codegen changes, actual changes are: https://github.com/openmsupply/open-msupply/compare/604-tests-for-sync-status...c955dbcb943f256470b69720dce1aa2c32f5c283, + 1.6k - 1.3k. Sorry I did want to make stacked PR's but it would take a bit longer, just wanted to get this out there to check.

## Changes

* Remove bootstrap and stages from server, just add swappable schema
* Add mutations: initialiseSite, manualSync, updateSyncSettings (vs just using the later for both initialise and update sync settings)
* Add queries: syncStatus, syncState (available without auth all the time), 
* Use sync log vs key value sore for syncStates
* If sync settings are changes check against central server
* Replace sync actor with sync driver (allows for either awaiting manualSync trigger or sync interval_sec)
* Add trigger to swap schema on sync first initialised
* Add sync progress component to front end
* Use new sync queries/mutations in initialisation form/phase in front end
* More comments

Sorry might get compilation errors but `yarn start-local` still works. See fix compilation error TODO will do this shortly.

## Explanation

Server now has graphql schema that is swapped after initialisation. 
syncState query will result in Initialised, Initialising and PreInitialisation, this is used by front end to go to login/initialisation page (it's available without auth for initialisation and operational schema).

Initialisation page check syncState and allows entering sync credentials in in PreInitialisation mode, if we are in Initialising mode sync credentials are not editable (initialise button is swapped to retry button if Initialising and there was an error, which allows triggering retry of sync via manual sync) <- if you start initialisation and refresh page, correct state should persist, same if you have entered settings via yaml and start the server and open f/e while server is initialising.

Added SyncProgress component that can also be used in manual sync in operational mode

## Test

(I found this command useful `rm -rf omsupply-database* && cargo run`)

- [ ] With fresh database, start server and f/e
- [ ] Should see sync progress
- [ ] Enter credentials and press initialise
- [ ] Should see progress
- [ ] Should be navigated to login page on completion
- [ ] Stop server, remove database and `cargo run`
- [ ] Add something to throw an error in 4d `syncV5API_queuedRecordsGet`
- [ ] Try initialise (should still error)
- [ ] Retry button should become visible and enabled while retrying
- [ ] Try retrying (prepare initial should not fire again (can see by the speed of that step)
- [ ] Try refresh page when error happens (or while initialising) state should persist
- [ ] Try clearing database again and entering settings in yaml file (still with error is in 4d method)
- [ ] Open front end and you should see the correct state
- [ ] Remove error in 4D
- [ ] Press retry again (sync should successfully complete and you should be navigated to login)

TODO

* Diagrams and documentation
* Fix compilation errors in update settings front end (I wanted to get this PR in asap, but only realised that there are some compilation errors in sync settings of operational mode in front end, will fix in another PR shortly)
* Add issues for moving `parseSyncErrorMessage` to a common place, and maybe bugsnagging errors that are not know, and returning standard graphql error rather then using regex
* if url is not correct we just get update sync settings failed error (only get validation error from server)
* Make sure back end tests still pass
